packages/trie/src/proof/range.ts

```diff
@@ -520,3 +701,121 @@ export async function verifyRangeProof(
 
   return hasRightElement(trie, keys[keys.length - 1])
 }
+
+/**
+ * Return a range proof of trie items between startingHash and limitHash
+ * This proof can be verified using `verifyRangeProof`.
+ * Note, that the proof will always include a proof for the startingHash (even if it does not exist)
+ * The limitHash could not be proved,
+ * @param startingHash This is the starting (hash) key item (note on non-hashed tries this is just a key)
+ * @param limitHash This is the limit (hash) key item (note on non-hashed tries this is just a key)
+ */
+export async function createRangeProof(
+  this: Trie,
+  startingKey: Uint8Array,
+  limitKey: Uint8Array
+): Promise<{
+  values: Uint8Array[]
+  keys: Uint8Array[]
+  proof: Uint8Array[] | null
+}> {
+  if (nibblesCompare(bytesToNibbles(startingKey), bytesToNibbles(limitKey)) > 0) {
+    throw new Error('invalid keys order')
+  }
+  let startingNode = (await this.findPath(startingKey)).node
+  let limitNode = (await this.findPath(limitKey)).node
+
+  if (
+    startingNode === null &&
+    limitNode === null &&
+    equalsBytes(startingKey, hexToBytes('0x' + '00'.repeat(32))) &&
+    equalsBytes(limitKey, hexToBytes('0x' + 'ff'.repeat(32)))
+  ) {
+    const keys: Uint8Array[] = []
+    const values: Uint8Array[] = []
+    const proof = null
+    await this.walkAllValueNodes(async (node, key) => {
+      keys.push(nibblestoBytes([...key, ...(node as any)._nibbles]))
+      values.push(node.value()!)
+    })
+    return { keys, values, proof }
+  }
+
+  const proofSet: Set<string> = new Set()
+  const keyvals: Map<string, Uint8Array> = new Map()
+
+  if (startingNode === null) {
+    // If the starting node does not exist, find the next value node to the right
+    const rightNode = await returnRightNode.bind(this)(bytesToNibbles(startingKey))
+    if (rightNode !== null) {
+      proofSet.add(bytesToHex(rightNode.node.serialize()))
+      startingKey = nibblestoBytes([...rightNode.currentKey, ...(rightNode.node as any)._nibbles])
+
+      keyvals.set(bytesToHex(startingKey), rightNode.node.value()!)
+      startingNode = rightNode.node
+    } else {
+      //
+    }
+  }
+
+  const limitNodeRight = await returnRightNode.bind(this)(bytesToNibbles(limitKey))
+  if (startingNode !== null && limitNodeRight === null) {
+    const walk = this.walkTrieIterable(this.root())
+    for await (const { node, currentKey } of walk) {
+      if (isValueNode(node)) {
+        const nodeKey = [...currentKey, ...(<any>node)._nibbles]
+        if (nibblesCompare(nodeKey, bytesToNibbles(startingKey)) >= 0) {
+          keyvals.set(bytesToHex(nibblestoBytes(nodeKey)), node._value!)
+          const path = await this.findPath(nibblestoBytes(nodeKey))
+          for (const p of path.stack) {
+            proofSet.add(bytesToHex(p.serialize()))
```

Comment:
I'm not sure, what does this do? Prove the start key? I think we only need the key of the rightmost key (so in this cast, left of `limitKey`)

Reply:
Do you mean that all we actually need to return is a proof for the limit-node?
Yes that would be much smaller....and make a lot more sense

Reply:
Yes, only the proof of the limitHash should be included in the range proof